File: typescript/design-as-you-code/domain/Car.ts
import Logger from "../shared/Logger";
import CarAirConditioner from "./CarAirConditioner";
import CarDoor from "./CarDoor";
import CarKey from "./CarKey";
import CarWaterSystem from "./CarWaterSystem";
import CarWindshieldWiper from "./CarWindshieldWiper";

export default class Car {
    private _locked = true;
    private _turnedOn = false;
    private _moving = false;
    private _doors = [new CarDoor(), new CarDoor(), new CarDoor(), new CarDoor()];
    private _windshieldWiper = new CarWindshieldWiper(this);
    private _waterSystem = new CarWaterSystem(this);
    private _airConditioner = new CarAirConditioner(this);

    constructor() {
        Logger.log('Car created');
    }

    get locked(): boolean {
        return this._locked;
    }

    get turnedOn(): boolean {
        return this._turnedOn;
    }

    get moving(): boolean {
        return this._moving;
    }

    get doors(): CarDoor[] {
        return [...this._doors];
    }

    get windshieldWiper(): CarWindshieldWiper {
        return this._windshieldWiper;
    }

    get waterSystem(): CarWaterSystem {
        return this._waterSystem;
    }
    
    get airConditioner(): CarAirConditioner {
        return this._airConditioner;
    }

    unlock(key: CarKey) {
        if(!key.authorizedFor(this)){
            return;
        }

        this._locked = false;

        Logger.log('Car unlocked');
    }

    openDoor(doorIndex: number) {
        if(this._locked) {
            return;
        }

        const selectedDoor = this._doors[doorIndex];

        if(!selectedDoor){
            return;
        }

        selectedDoor.open();

        Logger.log(`Door ${doorIndex} open`);
    }

    closeDoor(doorIndex: number) {
        if(this._locked) {
            return;
        }

        const selectedDoor = this._doors[doorIndex];

        if(!selectedDoor){
            return;
        }

        selectedDoor.close();

        Logger.log(`Door ${doorIndex} closed`);
    }

    turnOn(key: CarKey) {
        if(!key.authorizedFor(this)){
            return;
        }

        this._turnedOn = true;

        Logger.log('Car turned on');
    }

    turnOff(key: CarKey) {
        if(!key.authorizedFor(this)){
            return;
        }

        this._turnedOn = false;

        Logger.log('Car turned off');
    }
}

File: python/src/domain/Car.py
from src.shared.logger import log

from .carDoor import CarDoor
from .carWindshieldWiper import CarWindshieldWiper
from .carWaterSystem import CarWaterSystem
from .carAirConditioner import CarAirConditioner

class Car:
    def __init__(self):
        self._locked = True
        self._turnedOn = False
        self._moving = False
        self._doors = (CarDoor(), CarDoor(), CarDoor(), CarDoor())
        self._windshieldWiper = CarWindshieldWiper(self)
        self._waterSystem = CarWaterSystem(self)
        self._airConditioner = CarAirConditioner(self)

        log('Car created')

    @property
    def locked(self):
        return self._locked
    
    @property
    def turnedOn(self):
        return self._turnedOn
    
    @property
    def moving(self):
        return self._moving
    
    @property
    def doors(self):
        return self._doors
    
    @property
    def windshieldWiper(self):
        return self._windshieldWiper
    
    @property
    def waterSystem(self):
        return self._waterSystem
    
    @property
    def airConditioner(self):
        return self._airConditioner
    
    def unlock(self, carKey):
        if not carKey.authorizedFor(self):
            return

        self._locked = False

        log('Car unlocked')
    
    def openDoor(self, doorNumber):
        if self._locked:
            return

        self._doors[doorNumber].open()

        log(f'Door {doorNumber} open')

    def closeDoor(self, doorNumber):
        if self._locked:
            return

        self._doors[doorNumber].close()

        log(f'Door {doorNumber} closed')

    def turnOn(self, carKey):
        if not carKey.authorizedFor(self):
            return
        
        self._turnedOn = True

        log('Car turned on')

    def turnOff(self, carKey):
        if not carKey.authorizedFor(self):
            return
        
        self._turnedOn = False

        log('Car turned off')